### PR TITLE
Issue #50636: Improve error diagnostic with missing commas after struct fields.

### LINKED ISSUE #50636: parse error recovery could be better with missing commas after struct fields
Test case: [(playground link)](https://play.rust-lang.org/?gist=67235946f5140d25364ffacba39cd3b0&version=nightly&mode=debug)
```rust
struct S {
    foo: u32 // <- missing comma
    bar: u32
}

fn main() {
    let s = S { foo: 5, bar: 6 };
}
```

Actual result (nightly):
```
error: expected `,`, or `}`, found `bar`
 --> src/main.rs:3:5
  |
3 |     bar: u32
  |     ^^^
  |
  = help: struct fields should be separated by commas

error[E0560]: struct `S` has no field named `foo`
 --> src/main.rs:7:17
  |
7 |     let s = S { foo: 5, bar: 6 };
  |                 ^^^ `S` does not have this field

error[E0560]: struct `S` has no field named `bar`
 --> src/main.rs:7:25
  |
7 |     let s = S { foo: 5, bar: 6 };
  |                         ^^^ `S` does not have this field
```

Expected result:

Ideally the parser would guess that there was meant to be a comma at the end of the line.  Then:

- The "expected comma" error could highlight the line missing a comma (and suggest a fix-it), rather than highlighting the first token in the next line.
- After producing an error, it could recover by pretending the comma was there, adding the `foo` and `bar` fields to the struct to avoid the latter two errors.

### PR DESCRIPTION
**Pull request by simartin**
Fixes #50636

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -5797,9 +5797,18 @@ impl<'a> Parser<'a> {
...
-            _ => return Err(self.span_fatal_help(self.span,
-                    &format!("expected `,`, or `}}`, found `{}`", self.this_token_to_string()),
-                    "struct fields should be separated by commas")),
+            _ => {
+                let sp = self.sess.codemap().next_point(self.prev_span);
+                let mut err = self.struct_span_err(sp, &format!("expected `,`, or `}}`, found `{}`",
+                                                                self.this_token_to_string()));
+                if self.token.is_ident() {
+                    // This is likely another field; emit the diagnostic and keep going
+                    err.span_suggestion(sp, "try adding a comma", "".into());
```
Comment (oli-obk) [inline src/libsyntax/parse/parser.rs]: There should be a comma inside the `""` ;)

**Review by oli-obk — CHANGES_REQUESTED** [review]
Can you make the test a UI test? That makes it simpler to figure out whether the diagnostics show up nicely

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (simartin) [reply src/libsyntax/parse/parser.rs]: Thanks for the review! PR amended accordingly.

**Review by oli-obk — APPROVED** [review]

**Pull request merged**